File: UVT/Models/CopperPair.swift
//
//  CopperPair.swift
//  UVT
//
//  Created by Justin Kumpe on 8/2/22.
//

import UIKit

struct CopperPair {
    let pair: Int!
    let basePair: Int!
    let binder: Int!
    let superBinder: Int!
    let tip: String!
    let ring: String!
    let binderTip: String!
    let binderRing: String!
    let superBinderTip: String!
    let superBinderRing: String!
    let tipColor: UIColor!
    let ringColor: UIColor!
    let binderRingColor: UIColor!
    let binderTipColor: UIColor!
    let superBinderRingColor: UIColor!
    let superBinderTipColor: UIColor!
    init(pair: Int) {
        self.pair = pair
        var superBinder: Int {
            switch pair {
            case 1...625:
                return 1
            case 626...1250:
                return 2
            case 1251...1875:
                return 3
            case 1876...2500:
                return 4
            case 2501...3125:
                return 5
            case 3126...3750:
                return 6
            case 3751...4375:
                return 7
            case 4376...5000:
                return 8
            case 5001...5625:
                return 9
            case 5626...6250:
                return 10
            case 6251...6875:
                return 11
            case 6876...7500:
                return 12
            case 7501...8125:
                return 13
            case 8126...8750:
                return 14
            case 8751...9375:
                return 15
            case 9376...10000:
                return 16
            case 10001...10625:
                return 17
            case 10626...11250:
                return 18
            case 112551...11875:
                return 19
            case 11876...12500:
                return 20
            case 12501...13125:
                return 21
            case 13126...13750:
                return 22
            case 13751...14375:
                return 23
            case 14376...15000:
                return 24
            case 15001...15625:
                return 25
            default:
                return 0
            }
        }

        self.superBinder = superBinder
        let superBinderSub = (superBinder - 1) * 625
        let baseBinderPair = pair - superBinderSub

        var binder: Int {
            switch baseBinderPair {
            case 1...25:
                return 1
            case 26...50:
                return 2
            case 51...75:
                return 3
            case 76...100:
                return 4
            case 101...125:
                return 5
            case 126...150:
                return 6
            case 151...175:
                return 7
            case 176...200:
                return 8
            case 201...225:
                return 9
            case 226...250:
                return 10
            case 251...275:
                return 11
            case 276...300:
                return 12
            case 301...325:
                return 13
            case 326...350:
                return 14
            case 351...375:
                return 15
            case 376...400:
                return 16
            case 401...425:
                return 17
            case 426...450:
                return 18
            case 451...475:
                return 19
            case 476...500:
                return 20
            case 501...525:
                return 21
            case 526...550:
                return 22
            case 551...575:
                return 23
            case 576...600:
                return 24
            case 601...625:
                return 25
            default:
                return 0
            }
        }
        self.binder = binder
        let binderSub = (binder - 1) * 25
        let basePair = baseBinderPair - binderSub
        self.basePair = basePair
        self.tip = getTip(basePair)
        self.ring = getRing(basePair)
        self.binderTip = getTip(self.binder)
        self.binderRing = getRing(self.binder)
        self.superBinderTip = getTip(self.superBinder)
        self.superBinderRing = getRing(self.superBinder)
        self.tipColor = getColor(self.tip)
        self.ringColor = getColor(self.ring)
        self.binderTipColor = getColor(self.binderTip)
        self.binderRingColor = getColor(self.binderRing)
        self.superBinderTipColor = getColor(self.superBinderTip)
        self.superBinderRingColor = getColor(self.superBinderRing)
    }
}

private func getTip(_ pair: Int) -> String {
    switch pair {
    case 1...5:
        return "white"
    case 6...10:
        return "red"
    case 11...15:
        return "black"
    case 16...20:
        return "yellow"
    case 21...25:
        return "violet"
    default:
        return "unknown"
    }
}

private func getRing(_ pair: Int) -> String {
    switch pair {
    case 1, 6, 11, 16, 21:
        return "blue"
    case 2, 7, 12, 17, 22:
        return "orange"
    case 3, 8, 13, 18, 23:
        return "green"
    case 4, 9, 14, 19, 24:
        return "brown"
    case 5, 10, 15, 20, 25:
        return "slate"
    default:
        return "unknown"
    }
}

private func getColor(_ color: String) -> UIColor {
    switch color {
    case "blue":
        return .blue
    case "orange":
        return .orange
    case "green":
        return .green
    case "brown":
        return .brown
    case "slate":
        return .gray
    case "white":
        return .white
    case "red":
        return .red
    case "black":
        return .black
    case "yellow":
        return .yellow
    case "violet":
        return .purple
    default:
        return .clear
    }
}

File: classes/CopperPair.py
"""Copper Pair Model Class"""  # pylint: disable=invalid-name


class CopperPair:
    """Copper Pair"""

    def __init__(self, pair: int):
        self.pair = pair

    @staticmethod
    def __tip(pair) -> str:
        """Get Tip Color"""
        match pair:
            case num if 1 <= num <= 5:
                return "white"
            case num if 6 <= num <= 10:
                return "red"
            case num if 11 <= num <= 15:
                return "black"
            case num if 16 <= num <= 20:
                return "yellow"
            case num if 21 <= num <= 25:
                return "violet"
            case _:
                return "unknown"

    @property
    def tip_color(self) -> str:
        """Tip Color"""
        return self.__tip(self.pair)

    @staticmethod
    def __tip_hex(tip: str) -> str:
        """Get Tip Color HEX Code"""
        match tip:
            case "white":
                return "#ffffff"
            case "red":
                return "#ff0000"
            case "black":
                return "#000000"
            case "yellow":
                return "#ffff00"
            case "violet":
                return "#ee82ee"
            case _:
                return ""

    @property
    def tip_hex(self) -> str:
        """Tip Color HEX Code"""
        return self.__tip_hex(self.tip_color)

    @staticmethod
    def __ring(pair) -> str:
        """Get Ring Color"""
        match pair:
            case 1, 6, 11, 16, 21:
                return "blue"
            case 2, 7, 12, 17, 22:
                return "orange"
            case 3, 8, 13, 18, 23:
                return "green"
            case 4, 9, 14, 19, 24:
                return "brown"
            case 5, 10, 15, 20, 25:
                return "slate"
            case _:
                return "unknown"

    @property
    def ring_color(self) -> str:
        """Ring Color"""
        return self.__ring(self.pair)

    @staticmethod
    def __ring_hex(tip: str) -> str:
        """Get Ring Color HEX Code"""
        match tip:
            case "blue":
                return "#0000ff"
            case "orange":
                return "#ffa500"
            case "green":
                return "#008000"
            case "brown":
                return "#783f04"
            case "slate":
                return "#999999"
            case _:
                return ""

    @property
    def ring_hex(self) -> str:
        """Ring Color HEX Code"""
        return self.__ring_hex(self.ring_color)

    @property
    def super_binder(self) -> int:
        """Super Binder Group"""
        match self.pair:
            case num if 1 <= num >= 625:
                return 1
            case num if 626 <= num >= 1250:
                return 2
            case num if 1251 <= num >= 1875:
                return 3
            case num if 1876 <= num >= 2500:
                return 4
            case num if 2501 <= num >= 3125:
                return 5
            case num if 3126 <= num >= 3750:
                return 6
            case num if 3751 <= num >= 4375:
                return 7
            case num if 4376 <= num >= 5000:
                return 8
            case num if 5001 <= num >= 5625:
                return 9
            case num if 5626 <= num >= 6250:
                return 10
            case num if 6251 <= num >= 6875:
                return 11
            case num if 6876 <= num >= 7500:
                return 12
            case num if 7501 <= num >= 8125:
                return 13
            case num if 8126 <= num >= 8750:
                return 14
            case num if 8751 <= num >= 9375:
                return 15
            case num if 9376 <= num >= 10000:
                return 16
            case num if 10001 <= num >= 10625:
                return 17
            case num if 10626 <= num >= 11250:
                return 18
            case num if 112551 <= num >= 11875:
                return 19
            case num if 11876 <= num >= 12500:
                return 20
            case num if 12501 <= num >= 13125:
                return 21
            case num if 13126 <= num >= 13750:
                return 22
            case num if 13751 <= num >= 14375:
                return 23
            case num if 14376 <= num >= 15000:
                return 24
            case num if 15001 <= num >= 15625:
                return 25
            case _:
                return 0

    @property
    def binder(self) -> int:
        """Binder Number"""
        match self.base_binder_pair:
            case num if 1 <= num >= 25:
                return 1
            case num if 26 <= num >= 50:
                return 2
            case num if 51 <= num >= 75:
                return 3
            case num if 76 <= num >= 100:
                return 4
            case num if 101 <= num >= 125:
                return 5
            case num if 126 <= num >= 150:
                return 6
            case num if 151 <= num >= 175:
                return 7
            case num if 176 <= num >= 200:
                return 8
            case num if 201 <= num >= 225:
                return 9
            case num if 226 <= num >= 250:
                return 10
            case num if 251 <= num >= 275:
                return 11
            case num if 276 <= num >= 300:
                return 12
            case num if 301 <= num >= 325:
                return 13
            case num if 326 <= num >= 350:
                return 14
            case num if 351 <= num >= 375:
                return 15
            case num if 376 <= num >= 400:
                return 16
            case num if 401 <= num >= 425:
                return 17
            case num if 426 <= num >= 450:
                return 18
            case num if 451 <= num >= 475:
                return 19
            case num if 476 <= num >= 500:
                return 20
            case num if 501 <= num >= 525:
                return 21
            case num if 526 <= num >= 550:
                return 22
            case num if 551 <= num >= 575:
                return 23
            case num if 576 <= num >= 600:
                return 24
            case num if 601 <= num >= 625:
                return 25
            case _:
                return 0

    @property
    def base_pair(self) -> int:
        """Base Pair"""
        sub = (self.binder - 1) * 25
        return self.base_binder_pair - sub

    @property
    def base_binder_pair(self) -> int:
        """Base Binder Pair"""
        sub = (self.super_binder - 1) * 625
        return self.pair - sub

    @property
    def binder_tip_color(self) -> str:
        """Binder Tip Color"""
        return self.__tip_hex(self.__tip(self.binder))

    @property
    def binder_ring_color(self) -> str:
        """Binder Ring Color"""
        return self.__ring_hex(self.__tip(self.binder))

    @property
    def super_binder_tip_color(self) -> str:
        """Super Binder Tip Color"""
        return self.__tip_hex(self.__tip(self.super_binder))

    @property
    def super_binder_ring_color(self) -> str:
        """Super Binder Ring Color"""
        return self.__ring_hex(self.__ring(self.super_binder))
